### PR TITLE
fix(desktop): stop polling unavailable collaboration authority

### PR DESCRIPTION
**Pull request by testikun**
## Summary

Stop repeated background collaboration inbox polling after a Runtime Host reports the expected `operation_unavailable` result.

- The main-process IPC handler remembers the negative result for the lifetime of the registered Host candidate and returns an empty inbox without querying the Host again.
- The preload remembers unavailable owner scopes by `hostId` and `targetEpoch`, so subsequent polls skip the IPC call; reconnect/identity changes naturally clear the cache.
- Session-scoped collaboration queries keep their existing error behavior.

## Design

This keeps the existing Runtime Host wire protocol unchanged: no `host.status` capability field, compatibility-epoch bump, target identity expansion, or Runtime Host API surface is required. A new candidate registration resets the main-process memo, and a new target epoch invalidates the preload scope cache.

## Verification

- Node 24 `node --test apps/desktop/dist/main/__tests__/runtime-host-collaboration-ipc-main.test.js` — 2 passed.
- `npm --workspace @maka/desktop run build:preload` — passed.
- Biome check passed for the three changed source/test files.
- Ablation: removing the main-process memo made the second-query regression fail; the memo was restored.
- Full Desktop main build is currently blocked by pre-existing mainline locale/type drift outside this PR.

Refs #4522

## AI use

- [x] Generative tooling made a substantive contribution

Tool(s) and scope: OpenAI Codex authored the implementation and regression test.

### CODE REVIEW

File: packages/runtime-host/src/protocol/index.ts [hunk 1]
```diff
@@ -100,7 +100,7 @@ export const RUNTIME_HOST_REGISTRATION_SCHEMA_VERSION = 1 as const;
 export const RUNTIME_HOST_PROTOCOL_VERSION = 0 as const;
 // Increment when the same protocol version no longer guarantees safe Client-Host
 // interoperability. Mismatches are rejected before domain commands are admitted.
-export const RUNTIME_HOST_COMPATIBILITY_EPOCH = 92 as const;
+export const RUNTIME_HOST_COMPATIBILITY_EPOCH = 93 as const;
```
Comment (Phoenix500526) [inline packages/runtime-host/src/protocol/index.ts]: This branch now conflicts with `main`, where the compatibility epoch has already advanced to 94. Because the new `host.status` field is emitted by the Host and older strict decoders do not accept unknown fields, please rebase and advance this to the next epoch (currently 95), with a matching `// 95:` rationale. Keeping either 93 or 94 after resolving the conflict would not describe the resulting wire shape safely.

File: apps/desktop/src/main/runtime-host-collaboration-ipc-main.ts [hunk 2]
```diff
@@ -91,10 +93,26 @@ export function registerRuntimeHostCollaborationIpc(
...
+    async (_event, sessionId: unknown) => {
+      try {
+        return await client.queryCollaborationTurnRequests(
+          sessionId === undefined ? undefined : requiredId(sessionId, 'Session'),
+        );
+      } catch (error) {
+        if (
+          sessionId === undefined &&
+          error instanceof RuntimeHostOperationError &&
+          error.code === 'operation_unavailable'
```
Comment (Phoenix500526) [inline apps/desktop/src/main/runtime-host-collaboration-ipc-main.ts]: `main` now has overlapping handling that also verifies `error.operation === 'collaboration.turn-request.query'`. When rebasing, please retain that narrower predicate and add the `authorityUnavailable` marker to the existing empty-inbox projection, rather than replacing it with the code-only check here. That avoids classifying an unrelated `operation_unavailable` failure as a missing collaboration authority.

File: apps/desktop/src/main/__tests__/runtime-host-turn-request-inbox-preload.test.ts [hunk 3]
```diff
@@ -20,7 +20,35 @@
...
+});
+
+test('skips an Owner Host that explicitly lacks collaboration authority', () => {
+  const scopes = selectRuntimeHostCollaborationScopes([
+    { hostId: 'local', collaborationAuthority: false },
+    { hostId: 'remote', collaborationAuthority: true },
+    { hostId: 'legacy' },
+  ]);
+
+  assert.deepEqual(scopes.map(({ hostId }) => hostId), ['remote', 'legacy']);
```
Comment (Phoenix500526) [inline apps/desktop/src/main/__tests__/runtime-host-turn-request-inbox-preload.test.ts]: Could we add one regression test across the actual polling/cache boundary? The important behavior is: an unknown legacy Host is queried once, the `authorityUnavailable` response records `false`, and a second `getPendingTurnRequests()` call does not invoke the IPC query again. The helper tests here prove the pieces independently, but they would still pass if the marker were not wired into the metadata used by the next poll.

**Review by Phoenix500526 — COMMENTED** [review]
The capability-based direction looks sound. I found three merge-readiness items below; I am leaving this as comments only, not an approval or merge request.

**Comment by testikun** [comment]
Addressed in b325755 (rebased onto the latest apache/main):

- Advanced RUNTIME_HOST_COMPATIBILITY_EPOCH to 99 and documented the new host.status wire-shape rationale.
- Kept the empty-inbox fallback limited to the background query, requiring the exact collaboration.turn-request.query operation plus operation_unavailable; scoped session queries still rethrow.
- Added a polling/cache-boundary regression test proving an unknown legacy Host is queried once, caches authorityUnavailable=false, and is skipped on the next getPendingTurnRequests() poll.

Biome checks and the focused polling/cache tests pass locally. The full workspace build remains affected by pre-existing cross-package type drift under the available Node/npm environment.

**Comment by testikun** [comment]
已处理并重新提交。

- 已合并最新 apache/main，解决合并冲突。
- 保留 collaboration authority 的缓存/轮询修复。
- Runtime Host compatibility epoch 更新为 106，避免与主线及其他开放协议变更冲突。
- 本地针对性构建与测试通过（Runtime Host protocol、Desktop collaboration IPC、Host manager、preload polling：全部通过）。
- 按约束未运行本机完整 Host 生命周期测试。

**Comment by Phoenix500526** [comment]
The user-visible error stream from #4522 no longer reproduces on current `main` after #4557. That PR added a precise main-process fallback for the background `collaboration.turn-request.query`: when the Host returns `operation_unavailable`, the IPC handler now projects an empty inbox instread of rejecting, so Electron no longer prints the error every two secs. 

The underlying polling behavior still remains on `main`, though. Desktop continues sending the query on every inbox interval. The unsupported result is simply handled quietly. This PR offers the more complete capability-bounary solution: new Hosts report whether collaboration authority is available, Desktop skips Hosts that explicity lack it, and legacy/unknown Hosts are probed once and remembered as unavailable while transient failures retain their retry semantics.

So #4557 has already resolve the noisy terminal symptom, while this PR would remove the unnecessary IPC and Runtime Host traffic that remains.  @Astro-Han Could you take a look when you have times?

File: apps/desktop/src/main/runtime-host-desktop-manager.ts [hunk 4]
```diff
@@ -983,6 +1002,13 @@ class RuntimeHostDesktopManagerImpl implements RuntimeHostDesktopManager {
         throw error;
       }
       if (result.kind === 'ready') {
+        const status = result.candidate.client.status;
```
Comment (Astro-Han) [inline apps/desktop/src/main/runtime-host-desktop-manager.ts]: P2: this awaits an extra `host.status` round trip, up to 5 seconds, before `target.hostId` is set and the ready state is published. Every connect and every reconnect pays it, on a path that has nothing to do with collaboration, and a Host that is connected but slow stalls readiness for the full timeout.

It also looks redundant. The connect path already performs a `host.status` (`client/wait-for-ready.ts:34`), and `connection.ts:566` hands that result to the `onHostStatus` observer, which this PR registers a few lines above at 987. So the same fact already arrives on its own, and this is a second acquisition path for it. Smallest fix: delete lines 1005 to 1011 and keep the observer.

P3 on line 1006: `typeof status === 'function'` guards a method that always exists on `DesktopRuntimeHostClient`. It is there to tolerate the fake candidate in `runtime-host-desktop-manager.test.ts`. Fix the harness rather than leaving a test-shaped guard in production code.

File: apps/desktop/src/main/runtime-host-desktop-manager.ts [hunk 5]
```diff
@@ -971,6 +984,12 @@ class RuntimeHostDesktopManagerImpl implements RuntimeHostDesktopManager {
             // replay one-shot join progress.
             onConnectionPhase: (phase) => onConnectionPhase?.(phase),
             ...(refreshPeerRoutes ? {} : { refreshPeerRoutes: false }),
+            onHostStatus: (status) => {
```
Comment (Astro-Han) [inline apps/desktop/src/main/runtime-host-desktop-manager.ts]: P2: this overrides `onHostStatus` unconditionally, after `...target.input`. Before, it was only present when a caller supplied one (`...(onHostStatus ? { onHostStatus } : {})`).

That has a side effect the PR does not mention: `connection.ts:833-839` returns early from `#resetLivenessCheck` when a status observer exists, so ongoing Session traffic no longer defers the liveness timer and `host.status` keeps firing on schedule at `DEFAULT_LIVENESS_INTERVAL_MS = 2_000` (`connection.ts:87`). An idle connection already probed at that rate, so the increase is bounded, but the direction is the opposite of this PR's premise: to remove one 2 second query, every connection including the ones that do support collaboration gets a 2 second probe that traffic can no longer defer.

Resolve this together with the ready-path probe below. Either keep a main-process memo and register no observer at all, or keep the observer, register it only while the capability is unknown, and drop the explicit probe.

File: apps/desktop/src/main/__tests__/runtime-host-turn-request-inbox-preload.test.ts [hunk 6]
```diff
@@ -20,7 +20,59 @@
...
+  retainRuntimeHostCollaborationAuthority,
+  selectRuntimeHostCollaborationScopes,
+} from '../../preload/runtime-host-turn-request-inbox.js';
+
+test('retains a learned unavailable capability when a legacy identity omits it', () => {
+  assert.equal(retainRuntimeHostCollaborationAuthority(undefined, false), false);
+  assert.equal(retainRuntimeHostCollaborationAuthority(true, false), true);
+});
+
+test('caches an unavailable legacy Host across polling calls', async () => {
```
Comment (Astro-Han) [inline apps/desktop/src/main/__tests__/runtime-host-turn-request-inbox-preload.test.ts]: P3: this does not cross the boundary @Phoenix500526 asked about. The cache here is a local `authority` variable and a local mark closure, not `runtimeHostMetadata` and `markRuntimeHostCollaborationUnavailable`. Replace the callback passed at `preload.ts:1418` with a no-op, or have it write the wrong scope key, and this test still passes, which is exactly the failure mode the request was meant to rule out.

Assert against the real pair instead: mark through `markRuntimeHostCollaborationUnavailable`, then read back through `runtimeHostMetadataFor` on the second poll.

File: apps/desktop/src/preload/runtime-host-turn-request-inbox.ts [hunk 7]
```diff
@@ -19,6 +19,45 @@
...
+
+export function retainRuntimeHostCollaborationAuthority(
+  observed: boolean | undefined,
+  previous: boolean | undefined,
+): boolean | undefined {
+  return observed ?? previous;
+}
+
+/** Hosts with an explicit negative capability cannot answer collaboration queries. */
+export function selectRuntimeHostCollaborationScopes<T extends RuntimeHostCollaborationScope>(
```
Comment (Astro-Han) [inline apps/desktop/src/preload/runtime-host-turn-request-inbox.ts]: P3: `selectRuntimeHostCollaborationScopes` has no production caller. A grep over `apps/desktop/src` finds only this definition and its test; the real filter is inlined at line 51. Either use it at line 51 or delete it with its test, so the exported surface does not carry a function that pins nothing.

File: apps/desktop/src/main/runtime-host-collaboration-ipc-main.ts [hunk 8]
```diff
@@ -101,7 +102,11 @@ export function registerRuntimeHostCollaborationIpc(
         return await client.queryCollaborationTurnRequests(requestedSessionId);
       } catch (error) {
         if (requestedSessionId === undefined && isCollaborationInboxUnavailable(error)) {
-          return { canRequestTurns: false, requests: [] };
+          return {
+            canRequestTurns: false,
+            requests: [],
+            authorityUnavailable: true,
```
Comment (Astro-Han) [inline apps/desktop/src/main/runtime-host-collaboration-ipc-main.ts]: P3: `authorityUnavailable` is a cross-process contract field, but it exists only as an ad-hoc intersection at each end, here via `satisfies CollaborationTurnRequestQueryResult & { authorityUnavailable: true }` and at `preload.ts:1404` via a cast. It is not declared in `bridge-contract.d.ts`, so renaming or dropping it on one side compiles clean on the other. Declare the query result shape once in the bridge contract and have both ends use it.

File: apps/desktop/src/main/runtime-host-desktop-manager.ts [hunk 9]
```diff
@@ -467,12 +472,20 @@ class RuntimeHostDesktopManagerImpl implements RuntimeHostDesktopManager {
...
+      ...(target.collaborationAuthority === undefined
+        ? {}
+        : { collaborationAuthority: target.collaborationAuthority }),
       ...(candidate ? { candidate } : {}),
     };
   }
 
   entries(): readonly RuntimeHostDesktopTargetState[] {
-    return [...this.#targets.values()].map((target) => target.state);
+    return [...this.#targets.values()].map((target) => ({
```
Comment (Astro-Han) [inline apps/desktop/src/main/runtime-host-desktop-manager.ts]: P3: this spreads the generation's `collaborationAuthority` over `target.state`, but the state objects already carry the field from the constructors at 1227 to 1266 and from `#projectState`. One of the two writers is dead. Pick the state as the single carrier and drop the override here, or drop it from the state variants and keep the projection in one place.

**Review by Astro-Han — COMMENTED** [review]
Reviewed at edaa9be3. No P0 or P1. The main question is not correctness, it is whether the trade is worth making, so I want to put that first.

P2, and it decides the rest: after #4557 landed, the user-visible failure in #4522 is gone. @Phoenix500526 says so above and the code agrees, since this branch's base already contains the empty-inbox projection and this PR only adds a field to its return value. What is left is an optimization: one `collaboration.turn-request.query` every 2 seconds per Owner Host (`use-turn-request-inbox.ts:32`) that is known to fail. Worth weighing against that: the same connection already sends a `host.status` liveness probe every 2 seconds (`connection.ts:87`), so what gets removed is one small request on a connection that is already talking at that cadence.

The price is a new wire capability and a compatibility epoch bump, which means every deployed Host has to upgrade and reconnect before this Desktop build will talk to it. Against that, there is a smaller solution that gets the same result: `runtime-host-collaboration-ipc-main.ts:95-114` already closes over a single `client` registered per target epoch, and already catches the exact error. Remember it there and short-circuit. That is one file, a handful of lines, naturally scoped per connection, and invalidated by reconnect for free, with nothing in `packages/runtime-host` and no epoch bump. If you also want the renderer-to-main IPC gone, keep the `authorityUnavailable` marker and the preload cache and still drop the protocol capability, the identity plumbing and the retention rule.

If the protocol capability is worth keeping for reasons beyond this poll, say so in the description, because as written a reader sees an epoch bump that invalidates every older Host in exchange for one empty query every 2 seconds.

One thing the capability derivation gets right, since it is not obvious: `host-kernel.ts:842` derives it from `accessAuthority`, which is the same object that serves `collaboration.turn-request.query` at `:786-792`, so a false negative that silently kills a working inbox is not possible. The learned-negative cache is also bounded correctly: the preload scope key is `hostId` plus `targetEpoch` (`preload.ts:276-278`), so a reconnect on a new epoch drops what was learned rather than blacklisting a Host forever.

On the previous round: comment 1 (epoch plus rationale) and comment 2 (keep the narrower `error.operation` predicate) are both addressed. Comment 3 is not, see inline.

CI: there is no workflow run at all for this head. `gh pr checks` reports none, `actions/runs?head_sha=edaa9be3` returns 0, and the commit has 0 check runs with a `pending` combined status. It is not `action_required`, there is simply no run. The green result you may be looking at is the previous head `5bb90b9d`. The current head, the `Merge apache/main into collab polling` commit, is unverified. I could not determine from the API why it did not trigger.

Mergeability: `CONFLICTING`, and the only conflicting file is `packages/runtime-host/src/protocol/index.ts`. `main` is at epoch 109, so this needs 110 with the rationale carried over. That constant will keep conflicting, which is not your fault, but it does mean the rebase should be the last step before you ask for CI.

Evidence boundary: read only at edaa9be3, no build, no test run, no Desktop launched. The redundancy claim in the ready-path comment rests on a static read of `wait-for-ready.ts:34` and `connection.ts:566`; I did not observe that call.

AI-assisted review: drafted with Maka.

**Review by Astro-Han — APPROVED** [review]
Reviewed at 7b1bda76. This is now the shape from the last round: no wire capability, no epoch, the memo lives in the IPC closure that is created per candidate connection (`runtime-host-desktop-candidate.ts:833`), so a reconnect resets it for free. The main-process test crosses the boundary @Phoenix500526 asked for: second query, zero client calls.

One P3, and it is a deletion. The preload cache (`preload.ts:276`, `:1468-1486`) and the `authorityUnavailable` marker are a second copy of the same fact, keyed by `hostId` + `targetEpoch`. Reconnect keeps the epoch, and the key is only dropped when the scope is observed non-ready between two 2-second polls; a reconnect that completes inside one poll gap to an upgraded Host leaves the preload skipping it until Desktop restart. The main-process memo alone has none of that, and removing the preload half also removes the ad-hoc `CollaborationTurnRequestQueryResult & { authorityUnavailable?: true }` intersection that is not declared in `bridge-contract.d.ts`. That is −15 lines and the PR is one file plus its test.

Approving as-is at 7b1bda76 with that noted; if you drop the preload cache, ping me and I will re-check the new head.

AI-assisted review: drafted with Maka; I verified the IPC registration lifetime and the scope-key invalidation myself.

**Comment by testikun** [comment]
Applied the requested P3 cleanup in b4e530fdb:

- removed the preload unavailable-scope cache and the ad-hoc authorityUnavailable result intersection
- kept the main-process per-candidate memo and its regression assertion that the client is queried once
- typecheck, main/preload builds, and the focused 2-test suite pass

@Astro-Han please re-check the new head.

**Review by jackwener — COMMENTED** [review]
Reviewed at `b4e530fdbe80fc1e502b19c96f48f7ebdcb7d0b0`. The implementation is sound; I found no runtime correctness issue.

The negative result is cached only inside the collaboration IPC registration for one Runtime Host candidate. A replacement candidate creates a fresh registration and probes again. Only a background query receiving the exact `collaboration.turn-request.query` / `operation_unavailable` error sets the memo, while Session-scoped queries keep their existing error behavior. The focused two-test suite passed locally, and an additional production-module probe confirmed both properties: repeated calls use one Host query within a registration, and a fresh registration can recover to an available result. The exact-head hosted `test` check passed. A synthetic merge with current `main` (`8c52a835c35bb3bd8a5b5512644e1fedcf5a02fc`) was clean (`34f0006820204a78f79f73207e218dd3329a178f`).

**[P2] Please refresh the PR record and AI attribution before merge.** The current body still says that preload caches unavailable scopes by `hostId` and `targetEpoch`, that a new target epoch invalidates that cache, and that three source/test files changed. Head `b4e530fdb` removed the preload cache and the final diff is only the main-process handler plus its test, so those statements now describe code that no longer exists. In addition, the body says OpenAI Codex authored the implementation and regression test, but commit `7b1bda761` introduced the current memo and its regression assertions without the `Generated-by: Codex` trailer required by `CONTRIBUTING.md`. Please update the description to document the single main-process memo and amend every AI-authored affected commit to carry the required trailer.

The code path itself is ready after those contribution-record fixes. Please ping me after the amended head is green so I can bind the final review to it.

---

> **Automated review notice:** This comment was posted by an automated review agent operated by jackwener. It is not an independent human review and does not replace one.

**Comment by Astro-Han** [comment]
Thanks for the cleanup! The current main-process memo looks right; I have no further code changes to request.

Just two finishing touches, consistent with the latest review: update the PR description to remove the old preload/hostId/epoch cache design, and add the required `Generated-by: Codex` trailer to commits containing substantive Codex contributions (including `7b1bda761`). Please retain those trailers when squashing.

AI-assisted review with Codex.
